core/src/main/scala/cats/instances/invariant.scala

```diff
@@ -0,0 +1,168 @@
+        f(fa.combineAll(bs.map(g)))
+    }
+
+    def pure[A](a: A): Monoid[A] = new Monoid[A] {
+      val empty = a
+      def combine(x: A, y: A): A = a
+      override def combineAll(as: TraversableOnce[A]): A = a
+    }
+  }
+
+  implicit val catsInvariantMonoidalBand: InvariantMonoidal[Band] = new InvariantMonoidal[Band] {
+    def product[A, B](fa: Band[A], fb: Band[B]): Band[(A, B)] = new Band[(A, B)] {
+      def combine(x: (A, B), y: (A, B)): (A, B) = fa.combine(x._1, y._1) -> fb.combine(x._2, y._2)
+    }
+
+    def imap[A, B](fa: Band[A])(f: A => B)(g: B => A): Band[B] = new Band[B] {
+      def combine(x: B, y: B): B = f(fa.combine(g(x), g(y)))
+    }
+
+    def pure[A](a: A): Band[A] = new Band[A] {
+      def combine(x: A, y: A): A = a
+    }
+  }
+
+  implicit val catsInvariantMonoidalSemilattice: InvariantMonoidal[Semilattice] = new InvariantMonoidal[Semilattice] {
+    def product[A, B](fa: Semilattice[A], fb: Semilattice[B]): Semilattice[(A, B)] = new Semilattice[(A, B)] {
+      def combine(x: (A, B), y: (A, B)): (A, B) = fa.combine(x._1, y._1) -> fb.combine(x._2, y._2)
+    }
+
+    def imap[A, B](fa: Semilattice[A])(f: A => B)(g: B => A): Semilattice[B] = new Semilattice[B] {
+      def combine(x: B, y: B): B = f(fa.combine(g(x), g(y)))
+    }
+
+    def pure[A](a: A): Semilattice[A] = new Semilattice[A] {
+      def combine(x: A, y: A): A = a
+    }
+  }
+
+  implicit val catsInvariantMonoidalCommutativeSemigroup: InvariantMonoidal[CommutativeSemigroup] = new InvariantMonoidal[CommutativeSemigroup] {
+    def product[A, B](fa: CommutativeSemigroup[A], fb: CommutativeSemigroup[B]): CommutativeSemigroup[(A, B)] = new CommutativeSemigroup[(A, B)] {
+      def combine(x: (A, B), y: (A, B)): (A, B) = fa.combine(x._1, y._1) -> fb.combine(x._2, y._2)
+    }
+
+    def imap[A, B](fa: CommutativeSemigroup[A])(f: A => B)(g: B => A): CommutativeSemigroup[B] = new CommutativeSemigroup[B] {
+      def combine(x: B, y: B): B = f(fa.combine(g(x), g(y)))
+    }
+
+    def pure[A](a: A): CommutativeSemigroup[A] = new CommutativeSemigroup[A] {
+      def combine(x: A, y: A): A = a
+    }
+  }
+
+  implicit val catsInvariantMonoidalCommutativeMonoid: InvariantMonoidal[CommutativeMonoid] = new InvariantMonoidal[CommutativeMonoid] {
+    def product[A, B](fa: CommutativeMonoid[A], fb: CommutativeMonoid[B]): CommutativeMonoid[(A, B)] = new CommutativeMonoid[(A, B)] {
+      val empty = fa.empty -> fb.empty
+      def combine(x: (A, B), y: (A, B)): (A, B) = fa.combine(x._1, y._1) -> fb.combine(x._2, y._2)
+    }
+
+    def imap[A, B](fa: CommutativeMonoid[A])(f: A => B)(g: B => A): CommutativeMonoid[B] = new CommutativeMonoid[B] {
+      val empty = f(fa.empty)
+      def combine(x: B, y: B): B = f(fa.combine(g(x), g(y)))
+    }
+
+    def pure[A](a: A): CommutativeMonoid[A] = new CommutativeMonoid[A] {
+      val empty = a
+      def combine(x: A, y: A): A = a
+      override def combineAll(as: TraversableOnce[A]): A = a
```

Comment:
can we only override `combineAllOption`? If we override only `combineAll`, then calls to `combineAllOption` still get the slow path. `combineAll` calls back to `combineAllOption` (combineAll should probably be final, actually).

Reply:
I did override in all of the instances, but can you comment on how this is useful? I really can't come up with a way 🤔